sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala

```diff
@@ -207,11 +220,15 @@ case class TakeOrderedAndProject(
   private val ord: InterpretedOrdering = new InterpretedOrdering(sortOrder, child.output)
 
   // TODO: remove @transient after figure out how to clean closure at InsertIntoHiveTable.
-  @transient private val projection = projectList.map(new InterpretedProjection(_, child.output))
+  @transient private val projection = projectList.map(UnsafeProjection.create(_, child.output))
 
   private def collectData(): Array[InternalRow] = {
     val data = child.execute().map(_.copy()).takeOrdered(limit)(ord)
-    projection.map(data.map(_)).getOrElse(data)
+    if (projection.isDefined) {
+      projection.map(p => data.map(p(_).copy().asInstanceOf[InternalRow])).get
```

Comment:
Seems it will be a problem without this copy(). HiveCompatibilitySuite will be failed.
    [info]   key	value
    [info]   !== HIVE - 5 row(s) ==   == CATALYST - 5 row(s) ==
    [info]   !0	val_0                 4	val_4
    [info]   !0	val_0                 4	val_4
    [info]   !0	val_0                 4	val_4
    [info]   !2	val_2                 4	val_4
    [info]    4	val_4                 4	val_4